nixos/modules/virtualisation/xen-dom0.nix

```diff
@@ -157,354 +299,593 @@ in
+                  visible = false;
+                  description = "Set `access-log-special-ops`.";
+                };
+              };
+              xenfs = {
+                kva = lib.options.mkOption {
+                  default = "/proc/xen/xsd_kva";
+                  example = cfg.store.settings.xenstored.xenfs.kva;
+                  type = lib.types.path;
+                  visible = false;
+                  description = ''
+                    Path to the Xen Store Daemon KVA location inside the XenFS pseudo-filesystem.
+                    While it is possible to alter this value, some drivers may be hardcoded to follow the default paths.
+                  '';
+                };
+                port = lib.options.mkOption {
+                  default = "/proc/xen/xsd_port";
+                  example = cfg.store.settings.xenstored.xenfs.port;
+                  type = lib.types.path;
+                  visible = false;
+                  description = ''
+                    Path to the Xen Store Daemon userspace port inside the XenFS pseudo-filesystem.
+                    While it is possible to alter this value, some drivers may be hardcoded to follow the default paths.
+                  '';
+                };
+              };
+            };
+            ringScanInterval = lib.options.mkOption {
+              default = 20;
+              example = 30;
+              type = lib.types.addCheck (
+                lib.types.int
+                // {
+                  name = "nonzeroInt";
+                  description = "nonzero signed integer, meaning !=0";
+                  descriptionClass = "nonRestrictiveClause";
+                }
+              ) (n: n != 0);
+              description = ''
+                Perodic scanning for all the rings as a safenet for lazy clients.
+                Define the interval in seconds; set to a negative integer to disable.
+              '';
+            };
+          };
+        };
+      };
+    };
   };
 
-  ###### implementation
+  ## Implementation ##
 
-  config = mkIf cfg.enable {
+  config = lib.modules.mkIf cfg.enable {
     assertions = [
       {
         assertion = pkgs.stdenv.isx86_64;
-        message = "Xen currently not supported on ${pkgs.stdenv.hostPlatform.system}";
+        message = "Xen is currently not supported on ${pkgs.stdenv.hostPlatform.system}.";
       }
       {
-        assertion = config.boot.loader.grub.enable && (config.boot.loader.grub.efiSupport == false);
-        message = "Xen currently does not support EFI boot";
+        assertion = config.boot.loader.systemd-boot.enable || (config.boot ? lanzaboote) && config.boot.lanzaboote.enable;
+        message = "Xen currently does not support EFI boot on bootloaders that are not systemd-boot or Lanzaboote.";
+      }
+      {
+        assertion = config.boot.loader.grub.enable -> !config.boot.loader.grub.enableEfi;
+        message = "Booting Xen on legacy BIOS systems has been deprecated.";
```

Comment:
According to the message, I think the correct condition should be
```nix
        assertion = config.boot.loader.grub.enable -> config.boot.loader.grub.enableEfi;
```

Reply:
I don't understand how it is supposed to work.

Is xen works on grub, but only when using MBR boot?

Reply:
Wait, grub.enable can't be true, because we already have this assertion for systemd:
```nix
        assertion = config.boot.loader.systemd-boot.enable || (config.boot ? lanzaboote) && config.boot.lanzaboote.enable;
```

Reply:
The GRUB builder for Xen is broken, and should not be used.

However, I also want the assertion to be smart enough to tell BIOS users that Xen is not supported on BIOS, and to tell GRUB-efi users to use `systemd-boot` / Lanzaboote.

Reply:
I think systemd-boot assertion is enough, you can include statement about legacy boot/grub in here

Reply:
> Wait, grub.enable can't be true, because we already have this assertion, that only works when systemd-boot is used:
> 
> ```nix
>         assertion = config.boot.loader.systemd-boot.enable || (config.boot ? lanzaboote) && config.boot.lanzaboote.enable;
> ```

Actually, I might just be overengineering things for no reason. By rephrasing the EFI assertion, it fits both use cases.